### ISSUE TITLE
0.223 broke powerpc64 build

### ISSUE THREAD

**Issue by pkubaj** [mamedev/mame#7102]
Building on FreeBSD/head on powerpc64.

0.223 includes asmjit, which doesn't support ppc architectures at all.

Building results in:
```
Compiling src/devices/cpu/drcbex64.cpp...
In file included from ../../../../../src/devices/cpu/drcbex64.cpp:174:
../../../../../src/devices/cpu/drcbex64.h:23:25: error: expected namespace name
using namespace asmjit::x86;
                ~~~~~~~~^
../../../../../src/devices/cpu/drcbex64.h:103:3: error: unknown type name 'Gp'
                Gp select_register(Gp defreg) const;
                ^
../../../../../src/devices/cpu/drcbex64.h:103:22: error: unknown type name 'Gp'
                Gp select_register(Gp defreg) const;
                                   ^
../../../../../src/devices/cpu/drcbex64.h:104:3: error: unknown type name 'Xmm'; did you mean 'Imm'?
                Xmm select_register(Xmm defreg) const;
                ^~~
                Imm
In file included from ../../../../../src/devices/cpu/drcbex64.cpp:174:
In file included from ../../../../../src/devices/cpu/drcbex64.h:20:
In file included from ../../../../../3rdparty/asmjit/src/asmjit/asmjit.h:27:
In file included from ../../../../../3rdparty/asmjit/src/asmjit/./core.h:2006:
In file included from ../../../../../3rdparty/asmjit/src/asmjit/./core/../core/arch.h:28:
../../../../../3rdparty/asmjit/src/asmjit/./core/operand.h:1222:7: note: 'Imm' declared here
class Imm : public Operand {
      ^
In file included from ../../../../../src/devices/cpu/drcbex64.cpp:174:
../../../../../src/devices/cpu/drcbex64.h:104:23: error: unknown type name 'Xmm'; did you mean 'Imm'?
                Xmm select_register(Xmm defreg) const;
                                    ^~~
                                    Imm
In file included from ../../../../../src/devices/cpu/drcbex64.cpp:174:
In file included from ../../../../../src/devices/cpu/drcbex64.h:20:
In file included from ../../../../../3rdparty/asmjit/src/asmjit/asmjit.h:27:
In file included from ../../../../../3rdparty/asmjit/src/asmjit/./core.h:2006:
In file included from ../../../../../3rdparty/asmjit/src/asmjit/./core/../core/arch.h:28:
../../../../../3rdparty/asmjit/src/asmjit/./core/operand.h:1222:7: note: 'Imm' declared here
class Imm : public Operand {
      ^
In file included from ../../../../../src/devices/cpu/drcbex64.cpp:174:
../../../../../src/devices/cpu/drcbex64.h:118:2: error: unknown type name 'Mem'
        Mem MABS(const void *ptr, const uint32_t size = 0) const { return Mem(rbp, offset_from_rbp(ptr), size); }
        ^
../../../../../src/devices/cpu/drcbex64.h:122:2: error: unknown type name 'Gp'
        Gp get_base_register_and_offset(Assembler &a, void *target, Gp const &reg, int32_t &offset);
        ^
../../../../../src/devices/cpu/drcbex64.h:122:34: error: unknown type name 'Assembler'
        Gp get_base_register_and_offset(Assembler &a, void *target, Gp const &reg, int32_t &offset);
                                        ^
../../../../../src/devices/cpu/drcbex64.h:122:62: error: unknown type name 'Gp'
        Gp get_base_register_and_offset(Assembler &a, void *target, Gp const &reg, int32_t &offset);
                                                                    ^
../../../../../src/devices/cpu/drcbex64.h:123:22: error: unknown type name 'Assembler'
        void smart_call_r64(Assembler &a, x86code *target, Gp const &reg);
                            ^
../../../../../src/devices/cpu/drcbex64.h:123:53: error: unknown type name 'Gp'
        void smart_call_r64(Assembler &a, x86code *target, Gp const &reg);
                                                           ^
../../../../../src/devices/cpu/drcbex64.h:124:22: error: unknown type name 'Assembler'
        void smart_call_m64(Assembler &a, x86code **target);
                            ^
../../../../../src/devices/cpu/drcbex64.h:130:17: error: unknown type name 'Assembler'
        void op_handle(Assembler &a, const uml::instruction &inst);
                       ^
../../../../../src/devices/cpu/drcbex64.h:131:15: error: unknown type name 'Assembler'
        void op_hash(Assembler &a, const uml::instruction &inst);
                     ^
../../../../../src/devices/cpu/drcbex64.h:132:16: error: unknown type name 'Assembler'
        void op_label(Assembler &a, const uml::instruction &inst);
                      ^
../../../../../src/devices/cpu/drcbex64.h:133:18: error: unknown type name 'Assembler'
        void op_comment(Assembler &a, const uml::instruction &inst);
                        ^
../../../../../src/devices/cpu/drcbex64.h:134:17: error: unknown type name 'Assembler'
        void op_mapvar(Assembler &a, const uml::instruction &inst);
                       ^
../../../../../src/devices/cpu/drcbex64.h:136:14: error: unknown type name 'Assembler'
        void op_nop(Assembler &a, const uml::instruction &inst);
                    ^
../../../../../src/devices/cpu/drcbex64.h:137:16: error: unknown type name 'Assembler'
        void op_debug(Assembler &a, const uml::instruction &inst);
                      ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

0.222 built fine.

**Comment by MooglyGuy** [comment]
To my knowledge, this has been fixed in our CI, which also targets s390x. PowerPC is not a currently-supported build platform for MAME, so my advice would be to look back at the commit history and adapt the s390x fix to PPC as well.

**Comment by pkubaj** [comment]
Thanks, I'm taking a look at `makefile`. I can already see that there are entries for alpha, riscv64 and mips64, which are probably much less popular on desktops (which I assume MAME targets) than ppc64.

**Comment by MooglyGuy** [comment]
You would be mistaken that MAME targets them. If you would care to do a `git blame` on the relevant lines, they were added over two years ago, presumably as part of @belegdol 's efforts to test more CI targets, nothing more.

The supported build targets for MAME are the ones found advertised on http://www.mamedev.org/ . I am leaving this issue open only out of the hope that Julian or someone sees that he was tagged and takes a look at the issue out of common courtesy, not because it is a valid issue for the MAME team to look at.

**Comment by pkubaj** [comment]
I'm testing a fix now, will report on that:
```
--- makefile.orig       2020-08-05 19:48:15 UTC
+++ makefile
@@ -172,6 +172,9 @@ endif
 ifneq ($(filter powerpc,$(UNAME_P)),)
 PLATFORM := powerpc
 endif
+ifneq ($(filter powerpc64,$(UNAME_P)),)
+PLATFORM := powerpc64
+endif
 ifneq ($(filter riscv64%,$(UNAME_M)),)
 PLATFORM := riscv64
 endif
@@ -316,6 +319,9 @@ endif
 ifeq ($(firstword $(filter amd64,$(UNAME))),amd64)
 ARCHITECTURE := _x64
 endif
+ifeq ($(firstword $(filter powerpc64,$(UNAME))),powerpc64)
+ARCHITECTURE := _x64
+endif
 ifeq ($(firstword $(filter ppc64,$(UNAME))),ppc64)
 ARCHITECTURE := _x64
 endif
@@ -370,6 +376,12 @@ ifndef NOASM
 endif
 endif

+ifeq ($(findstring powerpc64,$(UNAME)),powerpc64)
+ifndef NOASM
+       NOASM := 1
+endif
+endif
+
 ifeq ($(findstring s390x,$(UNAME)),s390x)
 ifndef NOASM
        NOASM := 1
```

**Comment by pkubaj** [comment]
Only the third hunk should be necessary, others are to make powerpc64 recognized.

**Comment by belegdol** [comment]
Please keep in mind that PPC build is likely to fail due to #3157. I would have enabled Travis for PPC too if it were not for this issue.
Other than that, the fix seems valid. The logic in the makefile was searching for ppc64, not for powerpc64. I am guessing this is Linux vs FreeBSD difference.

**Comment by belegdol** [comment]
Please also note that you might need to see if powerpc64 does not need to be added to BIGENDIAN logic around line 400 in the makefile.

**Comment by cuavas** [comment]
Both ppc64 and ppc64le worked at least as far as building and running `-valid` before we started requiring `constexpr long double` maths.  I got the RGB VMX utilities working for little endian PowerPC myself.  The changes to stop asmjit from being used on s390x should be enough to get PowerPC support back to the state it was in before asmjit was added.

MAME definitely doesn’t “target” RISC-V, MIPS, etc.  They’re very much “if it breaks, you get to keep the pieces” status.  We might be a little more interested in PowerPC support than these other targets, but until the GNU multi-precision maths library supports the `long double` format used on PowerPC there’s not much we can do.  That’s not a compiler shortcoming I’m prepared to work around.
